HLSPlugin/src/com/kaltura/hls/manifest/HLSManifestEncryptionKey.as

```diff
@@ -106,18 +107,76 @@ package com.kaltura.hls.manifest
 		 * Decrypts a video or audio stream using AES-128 with the provided initialization vector.
 		 */
 		
-		public function decrypt( data:ByteArray, iv:ByteArray ):void
+		public function decrypt( data:ByteArray, iv:ByteArray ):ByteArray
 		{
 			trace("got " + data.length + " bytes");
 			if(data.length == 0)
-				return;
+				return data;
 				
 			var startTime:uint = getTimer();
-			decryptAES( data, _keyData, iv );
-			if ( usePadding ) unpad( data );
+			_key = new FastAESKey(_keyData);
+			iv.position = 0;
+			iv0 = iv.readUnsignedInt();
+			iv1 = iv.readUnsignedInt();
+			iv2 = iv.readUnsignedInt();
+			iv3 = iv.readUnsignedInt();
+			data.position = 0;
+			data = _decryptCBC(data,data.length);
+			//decryptAES( data, _keyData, iv );
+			if ( usePadding ){
+				data = unpad( data );
+			}
 			trace( "DECRYPTION OF " + data.length + " BYTES TOOK " + ( getTimer() - startTime ) + " MS" );
+			return data;
 		}
 		
+		
+		/* Cypher Block Chaining Decryption, refer to
+		* http://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Cipher-block_chaining_
+		* for algorithm description
+		*/
+		private function _decryptCBC(crypt : ByteArray, len : uint) : ByteArray {
+			var src : Vector.<uint> = new Vector.<uint>(4);
+			var dst : Vector.<uint> = new Vector.<uint>(4);
+			var decrypt : ByteArray = new ByteArray();
+			decrypt.length = len;
+			
+			for (var i : uint = 0; i < len / 16; i++) {
+				// read src byte array
+				src[0] = crypt.readUnsignedInt();
+				src[1] = crypt.readUnsignedInt();
+				src[2] = crypt.readUnsignedInt();
+				src[3] = crypt.readUnsignedInt();
+				
+				// AES decrypt src vector into dst vector
+				_key.decrypt128(src, dst);
+				
+				// CBC : write output = XOR(decrypted,IV)
+				decrypt.writeUnsignedInt(dst[0] ^ iv0);
+				decrypt.writeUnsignedInt(dst[1] ^ iv1);
+				decrypt.writeUnsignedInt(dst[2] ^ iv2);
+				decrypt.writeUnsignedInt(dst[3] ^ iv3);
+				
+				// CBC : next IV = (input)
+				iv0 = src[0];
+				iv1 = src[1];
+				iv2 = src[2];
+				iv3 = src[3];
+			}
+			//decrypt.position = 0;
+			return decrypt;
+		}
+		public function unpad(a : ByteArray) : ByteArray {
```

Comment:
@bengarney  not sure why but when i got out of the function i saw the old value.